src/Server/HTTP/authenticateUserByHTTP.cpp

```diff
@@ -0,0 +1,269 @@
+    extern const int SUPPORT_IS_DISABLED;
+}
+
+
+namespace
+{
+    String base64Decode(const String & encoded)
+    {
+        String decoded;
+        Poco::MemoryInputStream istr(encoded.data(), encoded.size());
+        Poco::Base64Decoder decoder(istr);
+        Poco::StreamCopier::copyToString(decoder, decoded);
+        return decoded;
+    }
+
+    String base64Encode(const String & decoded)
+    {
+        std::ostringstream ostr; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+        ostr.exceptions(std::ios::failbit);
+        Poco::Base64Encoder encoder(ostr);
+        encoder.rdbuf()->setLineLength(0);
+        encoder << decoded;
+        encoder.close();
+        return ostr.str();
+    }
+}
+
+
+bool authenticateUserByHTTP(
+    const HTTPServerRequest & request,
+    const HTMLForm & params,
+    HTTPServerResponse & response,
+    Session & session,
+    std::unique_ptr<Credentials> & request_credentials,
+    ContextPtr global_context,
+    LoggerPtr log)
+{
+    /// Get the credentials created by the previous call of authenticateUserByHTTP() while handling the previous HTTP request.
+    auto current_credentials = std::move(request_credentials);
+
+    /// The user and password can be passed by headers (similar to X-Auth-*),
+    /// which is used by load balancers to pass authentication information.
+    std::string user = request.get("X-ClickHouse-User", "");
+    std::string password = request.get("X-ClickHouse-Key", "");
+    std::string quota_key = request.get("X-ClickHouse-Quota", "");
+
+    /// The header 'X-ClickHouse-SSL-Certificate-Auth: on' enables checking the common name
+    /// extracted from the SSL certificate used for this connection instead of checking password.
+    bool has_ssl_certificate_auth = (request.get("X-ClickHouse-SSL-Certificate-Auth", "") == "on");
+    bool has_auth_headers = !user.empty() || !password.empty() || has_ssl_certificate_auth;
+
+    /// User name and password can be passed using HTTP Basic auth or query parameters
+    /// (both methods are insecure).
+    bool has_http_credentials = request.hasCredentials();
+    bool has_credentials_in_query_params = params.has("user") || params.has("password");
+
+    std::string spnego_challenge;
+    SSLCertificateSubjects certificate_subjects;
+
+    if (has_auth_headers)
+    {
+        /// It is prohibited to mix different authorization schemes.
+        if (has_http_credentials)
+            throw Exception(ErrorCodes::AUTHENTICATION_FAILED,
+                            "Invalid authentication: it is not allowed "
+                            "to use SSL certificate authentication and Authorization HTTP header simultaneously");
+        if (has_credentials_in_query_params)
+            throw Exception(ErrorCodes::AUTHENTICATION_FAILED,
+                            "Invalid authentication: it is not allowed "
+                            "to use SSL certificate authentication and authentication via parameters simultaneously simultaneously");
```

Comment:
in exception we mention SSL certificate but this triggers with any headers right not just SSL, right?

Reply:
The error messages were not correct, I've fixed them.